### PR TITLE
Ref bug #135533  Total of overtime is incorrect

### PR DESCRIPTION
**Pull request by nguyenminhduong**
Redmine https://dev.framgia.com/redmine/issues/135533
Total of overtime is incorrect when user creates OT request with time request is greater than 4 hours at the weekend and on Holiday

### CODE REVIEW

File: app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java [hunk 1]
```diff
@@ -157,9 +162,40 @@ public User getUser() {
...
+            return "";
+        }
+        Shifts shifts = new Shifts();
+        for (Branch branch : mUser.getBranches()) {
+            if (mRequestOverTime.getBranch().getBranchName().equals(branch.getBranchName())) {
+                shifts = branch.getShifts().get(0);
+            }
+        }
+        shifts.setTimeLunch(DateTimeUtils.convertUiFormatToDataFormat(shifts.getTimeLunch(),
+                DateTimeUtils.INPUT_TIME_FORMAT, DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM));
```
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.INPUT_TIME_FORMAT' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.INPUT_TIME_FORMAT'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

File: app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java [hunk 2]
```diff
@@ -157,9 +162,40 @@ public User getUser() {
...
+        shifts.setTimeAfternoon(DateTimeUtils.convertUiFormatToDataFormat(shifts.getTimeAfternoon(),
+                DateTimeUtils.INPUT_TIME_FORMAT, DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM));
```
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.INPUT_TIME_FORMAT' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.INPUT_TIME_FORMAT'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

File: app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java [hunk 3]
```diff
@@ -157,9 +162,40 @@ public User getUser() {
...
+        int hourOfTimeAfternoon = DateTimeUtils.getHourOfDay(shifts.getTimeAfternoon());
+        int minuteOfTimeAfternoon = DateTimeUtils.getMinute(shifts.getTimeAfternoon());
+        int hourOfTimeLunch = DateTimeUtils.getHourOfDay(shifts.getTimeLunch());
+        int minuteOfTimeLunch = DateTimeUtils.getMinute(shifts.getTimeLunch());
+        float numberHourOvertime = DateTimeUtils.getHourBetweenTwoDate(mRequestOverTime.getToTime(),
+                mRequestOverTime.getFromTime(), DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM);
```
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

File: app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java [hunk 4]
```diff
@@ -157,9 +162,40 @@ public User getUser() {
...
+        if (DateTimeUtils.checkHourOfDateLessThanOrEqual(mRequestOverTime.getFromTime(),
+                hourOfTimeLunch, minuteOfTimeLunch)
+                && !DateTimeUtils.checkHourOfDateLessThan(mRequestOverTime.getToTime(),
+                hourOfTimeAfternoon, minuteOfTimeAfternoon)
+                && DateTimeUtils.getHourBetweenTwoDate(mRequestOverTime.getToTime(),
+                mRequestOverTime.getFromTime(), DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM)
```
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

File: app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java [hunk 5]
```diff
@@ -157,9 +162,40 @@ public User getUser() {
...
+                > Constant.TimeConst.FOUR_HOUR) {
+            float numberHourLunch = DateTimeUtils.getHourBetweenTwoDate(shifts.getTimeAfternoon(),
+                    shifts.getTimeLunch(), DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM);
```
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

File: app/src/main/java/com/framgia/wsm/screen/managelistrequests/memberrequestdetail/MemberRequestDetailViewModel.java [hunk 6]
```diff
@@ -207,47 +211,76 @@ public String getCreateAtOffRequest() {
...
+        if (mOverTimeRequest.getFromTime() == null
+                || mOverTimeRequest.getToTime() == null
+                || mUser == null) {
             return "";
         }
-        return String.valueOf(DateTimeUtils.getHourBetweenTwoDate(getToTime(), getFromTime(),
-                DateTimeUtils.DATE_TIME_FORMAT_HH_MM_DD_MM_YYYY));
+        Shifts shifts = new Shifts();
+        for (Branch branch : mUser.getBranches()) {
+            if (mOverTimeRequest.getBranch().getBranchName().equals(branch.getBranchName())) {
```
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/managelistrequests/memberrequestdetail/MemberRequestDetailViewModel.java]: ERROR: [PMD-CPD] 13 lines duplicated.
app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java

File: app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java [hunk 7]
```diff
@@ -157,9 +162,40 @@ public User getUser() {
...
+    @Bindable
     public String getNumberHour() {
-        return String.valueOf(DateTimeUtils.getHourBetweenTwoDate(mRequestOverTime.getToTime(),
-                mRequestOverTime.getFromTime(), DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM));
+        if (mUser == null) {
+            return "";
+        }
+        Shifts shifts = new Shifts();
+        for (Branch branch : mUser.getBranches()) {
+            if (mRequestOverTime.getBranch().getBranchName().equals(branch.getBranchName())) {
```
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: ERROR: [PMD-CPD] 13 lines duplicated.
app/src/main/java/com/framgia/wsm/screen/managelistrequests/memberrequestdetail/MemberRequestDetailViewModel.java

File: app/src/main/java/com/framgia/wsm/screen/managelistrequests/memberrequestdetail/MemberRequestDetailViewModel.java [hunk 8]
```diff
@@ -207,47 +211,76 @@ public String getCreateAtOffRequest() {
...
+        int hourOfTimeLunch = DateTimeUtils.getHourOfDay(shifts.getTimeLunch());
+        int minuteOfTimeLunch = DateTimeUtils.getMinute(shifts.getTimeLunch());
+        float numberHourOvertime = DateTimeUtils.getHourBetweenTwoDate(getToTime(),
+                getFromTime(), DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM);
+        if (DateTimeUtils.checkHourOfDateLessThanOrEqual(getFromTime(), hourOfTimeLunch,
+                minuteOfTimeLunch)
+                && !DateTimeUtils.checkHourOfDateLessThan(getToTime(), hourOfTimeAfternoon,
+                minuteOfTimeAfternoon)
+                && DateTimeUtils.getHourBetweenTwoDate(getToTime(), getFromTime(),
+                DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM) > Constant.TimeConst.FOUR_HOUR) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/wsm/screen/managelistrequests/memberrequestdetail/MemberRequestDetailViewModel.java]: pls define a method ``isContainLunchBreak`` contain this conditions

File: app/src/main/java/com/framgia/wsm/screen/managelistrequests/memberrequestdetail/MemberRequestDetailViewModel.java [hunk 9]
```diff
@@ -207,47 +211,76 @@ public String getCreateAtOffRequest() {
...
+        }
+        shifts.setTimeLunch(DateTimeUtils.convertUiFormatToDataFormat(shifts.getTimeLunch(),
+                DateTimeUtils.INPUT_TIME_FORMAT, DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM));
+        shifts.setTimeAfternoon(DateTimeUtils.convertUiFormatToDataFormat(shifts.getTimeAfternoon(),
+                DateTimeUtils.INPUT_TIME_FORMAT, DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM));
+        int hourOfTimeAfternoon = DateTimeUtils.getHourOfDay(shifts.getTimeAfternoon());
+        int minuteOfTimeAfternoon = DateTimeUtils.getMinute(shifts.getTimeAfternoon());
+        int hourOfTimeLunch = DateTimeUtils.getHourOfDay(shifts.getTimeLunch());
+        int minuteOfTimeLunch = DateTimeUtils.getMinute(shifts.getTimeLunch());
+        float numberHourOvertime = DateTimeUtils.getHourBetweenTwoDate(getToTime(),
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/wsm/screen/managelistrequests/memberrequestdetail/MemberRequestDetailViewModel.java]: pls write a method ``getNumberHourOvertime`` inside ``Shift`` class and use it,
pls do same with other fields

File: app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java [hunk 10]
```diff
@@ -157,9 +162,42 @@ public User getUser() {
...
+                shifts = branch.getShifts().get(0);
+            }
+        }
+        float numberHourOvertime = DateTimeUtils.getHourBetweenTwoDate(mRequestOverTime.getToTime(),
+                mRequestOverTime.getFromTime(), DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM);
```
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

File: app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java [hunk 11]
```diff
@@ -157,9 +162,42 @@ public User getUser() {
...
+        if (isContainLunchBreak(shifts)) {
+            return String.format(Locale.getDefault(), DateTimeUtils.HOUR_FORMAT_2F,
+                    numberHourOvertime - shifts.getNumberHourLunch());
+        }
+        return String.format(Locale.getDefault(), DateTimeUtils.HOUR_FORMAT_2F, numberHourOvertime);
+    }
+
+    private boolean isContainLunchBreak(Shifts shifts) {
+        shifts.setTimeLunch(DateTimeUtils.convertUiFormatToDataFormat(shifts.getTimeLunch(),
+                DateTimeUtils.INPUT_TIME_FORMAT, DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM));
```
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.INPUT_TIME_FORMAT' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.INPUT_TIME_FORMAT'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

File: app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java [hunk 12]
```diff
@@ -157,9 +162,42 @@ public User getUser() {
...
+        shifts.setTimeAfternoon(DateTimeUtils.convertUiFormatToDataFormat(shifts.getTimeAfternoon(),
+                DateTimeUtils.INPUT_TIME_FORMAT, DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM));
```
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.DATE_TIME_FORMAT_YYYY_MM_DD_HH_MM'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'DateTimeUtils.INPUT_TIME_FORMAT' due to existing static import 'com.framgia.wsm.utils.common.DateTimeUtils.INPUT_TIME_FORMAT'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

File: app/src/main/java/com/framgia/wsm/screen/managelistrequests/memberrequestdetail/MemberRequestDetailViewModel.java [hunk 13]
```diff
@@ -207,47 +211,78 @@ public String getCreateAtOffRequest() {
...
         }
-        return String.valueOf(DateTimeUtils.getHourBetweenTwoDate(getToTime(), getFromTime(),
-                DateTimeUtils.DATE_TIME_FORMAT_HH_MM_DD_MM_YYYY));
+        Shifts shifts = new Shifts();
+        for (Branch branch : mUser.getBranches()) {
+            if (mOverTimeRequest.getBranch().getBranchName().equals(branch.getBranchName())) {
+                shifts = branch.getShifts().get(0);
+            }
+        }
+        float numberHourOvertime = DateTimeUtils.getHourBetweenTwoDate(getToTime(), getFromTime(),
```
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/managelistrequests/memberrequestdetail/MemberRequestDetailViewModel.java]: ERROR: [PMD-CPD] 19 lines duplicated.
app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java

File: app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java (hunk 10 above)
Comment (trainer-android-fr) [inline app/src/main/java/com/framgia/wsm/screen/requestovertime/confirmovertime/ConfirmOvertimeViewModel.java]: ERROR: [PMD-CPD] 19 lines duplicated.
app/src/main/java/com/framgia/wsm/screen/managelistrequests/memberrequestdetail/MemberRequestDetailViewModel.java

**Pull request merged**